okta/okta.go

```diff
@@ -468,67 +445,226 @@ func (m *userManager) GetGroupsForUser(user string) ([]string, error) {
+			"aud":   m.clientID,
+			"nonce": nonce,
+		},
+	}
+	return verifier.New().VerifyIdToken(token)
+}
+
+// validateAccessToken verifies that the access token is valid.
+// TODO (kim): figure out why this does not validate with the same jwt verifier
+// library as used for the ID token.
+func (m *userManager) validateAccessToken(token string) error {
+	info, err := m.getTokenInfo(context.Background(), token, "access_token")
+	if err != nil {
+		return errors.Wrap(err, "could not check if token is valid")
+	}
+	if !info.Active {
+		return errors.New("access token is inactive, so authorization is not possible")
+	}
+	return nil
+}
+
+// refreshTokens exchanges the given refresh token to redeem tokens from the
+// token endpoint.
+func (m *userManager) refreshTokens(ctx context.Context, refreshToken string) (*tokenResponse, error) {
+	q := url.Values{}
+	q.Set("grant_type", "refresh_token")
+	q.Set("refresh_token", refreshToken)
+	q.Set("scope", "openid email profile offline_access groups")
+	return m.redeemTokens(ctx, q.Encode())
 }
 
+// exchangeCodeForTokens exchanges the given code to redeem tokens from the
+// token endpoint.
+func (m *userManager) exchangeCodeForTokens(ctx context.Context, code string) (*tokenResponse, error) {
+	q := url.Values{}
+	q.Set("grant_type", "authorization_code")
+	q.Set("code", code)
+	q.Set("redirect_uri", m.redirectURI)
+	return m.redeemTokens(ctx, q.Encode())
+}
+
+// tokenResponse represents a response received from the token endpoint.
 type tokenResponse struct {
 	AccessToken      string `json:"access_token,omitempty"`
+	IDToken          string `json:"id_token,omitempty"`
+	RefreshToken     string `json:"refresh_token,omitempty"`
 	TokenType        string `json:"token_type,omitempty"`
 	ExpiresIn        int    `json:"expires_in,omitempty"`
 	Scope            string `json:"scope,omitempty"`
-	IDToken          string `json:"id_token,omitempty"`
-	RefreshToken     string `bson:"refresh_token,omitempty"`
 	ErrorCode        string `json:"error,omitempty"`
 	ErrorDescription string `json:"error_description,omitempty"`
 }
 
+// redeemTokens sends the request to redeem tokens with the required client
+// credentials.
+func (m *userManager) redeemTokens(ctx context.Context, query string) (*tokenResponse, error) {
+	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("%s/oauth2/v1/token", m.issuer), bytes.NewBufferString(query))
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	req = req.WithContext(ctx)
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	req.Header.Add("Accept", "application/json")
+	req.Header.Add("Connection", "close")
+	m.addAuthHeader(req)
+
+	client := m.getHTTPClient()
+	defer m.putHTTPClient(client)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	start := time.Now()
+	resp, err := client.Do(req)
+	grip.Info(message.Fields{
+		"endpoint":        "token",
+		"duration_millis": time.Since(start).Milliseconds(),
```

Comment:
This doesn't compile. But strangely enough the task passes (test fails though)

Reply:
Sorry, this was a go1.13 feature. I fixed it. I also fixed the makefile, which is still probably full of errors but probably fewer of them.